### DISCUSSION TITLE
Content Update Api fails when a newly created framework used with values other than  BMGS

### DISCUSSION THREAD

**Discussion by vinod303** [Sunbird-Knowlg/Community#137] [Bugs]
Found the following issue while creating a new framework category and use that in the content while saving.
SB version: 5.1
**STEPS TO REPLICATE**
1. Created Master categories called, "state","district", "subcategories" (Ref. https://knowlg.sunbird.org/learn/product-and-developer-guide/taxonomy-and-tagging/framework-service/faqs)
eg. 

`{
   "request": {
       "category": {
           "name": "State",
           "description": "State",
           "code": "state",
           "orgIdFieldName": "state",
           "targetIdFieldName": "state",
           "searchIdFieldName": "state",
           "searchLabelFieldName": "state"
       }
   }
}
`
2. Created a new framework eg. ka_agr_fw
3. Created  framework categories. eg
`{
  "request": {
     "category":{
         "name":"State",
         "description":"State",
         "code":"state"
     }
   }
}
`
4. Created Terms under the new category using the term creation api eg.
`{
   "request": {
       "term": {
           "name": "Karnataka",
           "label": "Karnataka",
           "description": "Karnataka",
           "code": "statekarnataka"
       }
   }
}
`
5. Updated the form to display these framework/framework values in the Content Creation form using the  form update api eg.
Here is the output of the form read
`"data": {
                "templateName": "defaultTemplate",
                "action": "save",
                "fields": [
                    {
                        "code": "appicon",
                        "visible": true,
                        "editable": true,
                        "dataType": "url",
                        "renderingHints": {},
                        "name": "App Icon",
                        "description": "App Icon",
                        "index": 1,
                        "inputType": "file",
                        "label": "Icon",
                        "placeholder": "App Icon",
                        "required": true
                    },
                    {
                        "code": "name",
                        "visible": true,
                        "editable": true,
                        "dataType": "text",
                        "renderingHints": {},
                        "name": "Title",
                        "description": "Title of the content",
                        "index": 2,
                        "inputType": "text",
                        "label": "Title",
                        "placeholder": "Enter Title For Book",
                        "required": true
                    },
                    {
                        "code": "description",
                        "visible": true,
                        "editable": true,
                        "dataType": "text",
                        "renderingHints": {},
                        "name": "Description",
                        "description": "Brief description",
                        "index": 3,
                        "inputType": "textarea",
                        "label": "Description",
                        "placeholder": "Brief description about the Book",
                        "required": false
                    },
                    {
                        "code": "keywords",
                        "visible": true,
                        "editable": true,
                        "dataType": "list",
                        "name": "Keywords",
                        "description": "Keywords for the content",
                        "index": 4,
                        "inputType": "keywordsuggestion",
                        "label": "keywords",
                        "placeholder": "Enter Keywords",
                        "required": false
                    },
                    {
                        "code": "state",
                        "visible": true,
                        "depends": [],
                        "editable": true,
                        "dataType": "text",
                        "renderingHints": {},
                        "description": "State",
                        "index": 5,
                        "label": "State",
                        "required": true,
                        "name": "State",
                        "inputType": "select",
                        "placeholder": "Select State"
                    },
                    {
                        "code": "district",
                        "visible": true,
                        "depends": [],
                        "editable": true,
                        "dataType": "list",
                        "renderingHints": {},
                        "description": "",
                        "index": 6,
                        "label": "District",
                        "required": true,
                        "name": "District",
                        "inputType": "multiselect",
                        "placeholder": "Select District"
                    },
                    {
                        "code": "subcategory",
                        "visible": true,
                        "depends": [],
                        "editable": true,
                        "dataType": "list",
                        "renderingHints": {},
                        "description": "",
                        "index": 8,
                        "label": "Subcategory",
                        "required": true,
                        "name": "Subcategory",
                        "inputType": "multiselect",
                        "placeholder": "Select Subcategory"
                    },
                    {
                        "code": "topic",
                        "visible": true,
                        "depends": [],
                        "editable": true,
                        "dataType": "list",
                        "renderingHints": {},
                        "description": "",
                        "index": 8,
                        "label": "Topic",
                        "required": true,
                        "name": "Topic",
                        "inputType": "multiselect",
                        "placeholder": "Select Topic"
                    },
                    {
                        "code": "audience",
                        "visible": true,
                        "editable": true,
                        "dataType": "list",
                        "renderingHints": {},
                        "description": "",
                        "index": 13,
                        "range": [
                            "Student",
                            "Teacher",
                            "Administrator",
                            "FLW",
                            "State Admin"
                        ],
                        "label": "Audience",
                        "required": false,
                        "name": "Audience",
                        "inputType": "select",
                        "placeholder": "Select Audience"
                    },
                    {
                        "code": "attributions",
                        "visible": true,
                        "editable": true,
                        "dataType": "list",
                        "renderingHints": {},
                        "name": "attribution",
                        "description": "Attributions",
                        "index": 15,
                        "inputType": "text",
                        "label": "Attributions",
                        "placeholder": "Attributions",
                        "required": false
                    },
                    {
                        "code": "copyright",
                        "visible": true,
                        "editable": true,
                        "dataType": "text",
                        "renderingHints": {},
                        "name": "Copyright",
                        "description": "Copyright",
                        "index": 22,
                        "inputType": "text",
                        "label": "Copyright",
                        "placeholder": "Copyright",
                        "required": false
                    },
                    {
                        "code": "license",
                        "visible": true,
                        "editable": true,
                        "defaultValue": "By creating and uploading content on DIKSHA, you consent to publishing this content under the Creative Commons Framework, specifically under the CC-BY-SA 4.0 license.",
                        "dataType": "text",
                        "renderingHints": {
                            "value": {
                                "video/x-youtube": "By linking or uploading YouTube videos on DIKSHA, you consent to publishing it as per the terms of the YouTube video license. DIKSHA accepts only videos with YouTube Standard License or Creative Commons License."
                            },
                            "class": "twelve wide field"
                        },
                        "description": "license",
                        "index": 25,
                        "label": "license",
                        "required": false,
                        "name": "license",
                        "inputType": "label",
                        "placeholder": "license"
                    }
                ]
            },`

6. When create a resource, the creation form appears with the new framework values
![image](https://github.com/Sunbird-Knowlg/Community/assets/15699608/5f36ec9f-dc9f-4a58-90a9-49c56413736c)
7. But, when I select values and then save it (eg. i selected "Karnataka" from "State"  ) and then save the content I get the following errors, which is returned by the Content Update API
`
{resmsgid: "f1bef740-0f18-11ee-beec-3d472e789049", msgid: null, status: "failed",…}
err
: 
"ERR_VALIDATING_CONTENT_FRAMEWORK"
errmsg
: 
"Nodes not found for Id's List(Bagalkot, Karnataka, Adult Health Management) "
msgid
: 
null
resmsgid
: 
"f1bef740-0f18-11ee-beec-3d472e789049"
status
: 
"failed"
responseCode
: 
"RESOURCE_NOT_FOUND"`

Can you please help and let me know what is causing these errors? Did i miss any steps? 

  @Krishnaj20 @pallakartheekreddy

**Comment by pallakartheekreddy** [comment]
@vinod303 Content service will maintain framework data in memory cache. Please restart content-service to get the latest framework data.

**Reply by vinod303** [reply]
@pallakartheekreddy  thanks!
Understood the issue,  The issue caused due to the wrong passing of rparameters at the time of master category creation. I initially used
`{ "request": { "category": { "name": "State", "description": "State", "code": "state", "orgIdFieldName": "state", "targetIdFieldName": "state", "searchIdFieldName": "state", "searchLabelFieldName": "state" } } } `

But , it is supposed to be like

`{
   "request": {
       "category": {
           "name": "State",
           "description": "State",
           "code": "state",
           "orgIdFieldName": "stateIds",
           "targetIdFieldName": "targetStateIds",
           "searchIdFieldName": "se_stateIds",
           "searchLabelFieldName": "se_states"
       }
   }
}`


ie. the orgIdFieldName,targetIdFieldName,searchIdFieldName, searchLabelFieldName  should be in a certain format.   

i will probably write a small how to and publish in this forum on how to create and use master categories.
